### DISCUSSION TITLE
Bottle merging in pr-pull not creating an "all" checksum for non-compiled formulae

### DISCUSSION THREAD

**Discussion by drbenmorgan** [Homebrew/discussions#3813] [Tap maintenance and Homebrew development]
I've been trying out the `brew tap-new` created GitHub Actions (this is a great feature, thanks!) to build and publish bottles for a library I'm working on. Several of the dependencies of the library are non-compiled data files, and thus will be identical across all macOS versions and Linux. The `install` block of these formulae are nothing more than a copy of the files from the unpacked area to the final location in the cellar. As far as I can tell, I'm doing exactly what's done in similar `homebrew-core` Formulae, e.g. `zsh-completions.rb`.

Bottles are built, merged and published without error, but I'm getting different checksums on the three builds I'm doing (via Github Runners), e.g.

```ruby
  bottle do
    sha256 cellar: :any_skip_relocation, monterey:     "4da9667d2f91ce11377bbe68ad652c7c6f37b9e9c49ecde2a6ef7d50b3573af3"
    sha256 cellar: :any_skip_relocation, big_sur:      "ef779219b2e52c06e92ec3d6885c6e7bd44c5c44dd08d81dd6e9fa9dce061052"
    sha256 cellar: :any_skip_relocation, x86_64_linux: "149254f26e243dffc99ca8d6cc31489682b987e93ff32eca69ebd380edc9211a"
  end
```

whilst I'd _expected_ to get a `sha256 cellar: :any_skip_reolcation, all: "COMMONHASH"` line from the bottle merging. Downloading the bottle artifacts, they do differ in:

- Timestamp of the `.tar.gz` bottle file
- Contents on the `INSTALL_RECEIPT.json` file inside these (in terms of the compiler/built_on).

Unpacking the bottles shows the copied in files from the install block are all identical with the same checksums, as expected. 

That the timestamp and `INSTALL_RECEIPT.json` differ isn't really a surprise given differing bottle creation time/platforms, but it seems like `homebrew-core` must solve this (though I'm aware that uses self-hosted runners). I'm therefore not sure whether bottle merging isn't expected to work/not supported on custom taps, there's a problem in my Formula (see https://github.com/drbenmorgan/homebrew-geant4/blob/main/Formula/g4ensdfstate.rb), or something's missing from the actions generated by `tap-new`.

NB: I did wonder about not bottling these particular formulae at all, but as `homebrew-core` now requires bottles for everything(?), I'd like to keep the tap following practices there!

**Comment by carlocab** [comment]
Pass `--only-json-tab` to `brew test-bot` here: https://github.com/drbenmorgan/homebrew-geant4/blob/057787f8b49a7001a2ef9c808d3e631c6f7aa4c8/.github/workflows/tests.yml#L36

**Comment by drbenmorgan** [comment]
Thanks @carlocab, that now makes the bottle checksums the same on both macOS versions, but the linux one is now different. Bottle contents are now identical, only clear difference is in timestamps of `.tar.gz` bottles, which differ by ~10seconds on the macOS bottles, but ~4mins between macOS and Linux. I'll have a bit more of a read on the brew Issues/code to see if there's anything related to bottle timestamps there.

**Reply by carlocab** [reply]
I don't think it's the timestamps that differ, but the permission bits: [out.html.gz](https://github.com/Homebrew/discussions/files/9833309/out.html.gz)

**Reply by drbenmorgan** [reply]
Thanks again @carlocab! This now looks basically the same as [another discussion](https://github.com/orgs/Homebrew/discussions/1640) which had an answer but not an immediate solution.

What I've tried was to break down the `brew test-bot` step in `tests.yml` as generated by `tap-new` and see where the differing permissions are introduced. This was/is extremely hacky, but what appears clear is that when running on a GitHub hosted runner for `ubuntu-22.04`, the  `g+w` permissions on the formula directories (as shown in your out file) are there _after_ `brew install` of the formula. On the macOS runners, the directories are not group writable at this same point.

I also tried running the Linux bottling workflow in Homebrew's `ghcr.io/homebrew/ubuntu22.04:master` image, and that *does* work and produce a linux bottle with identical checksum to the macOS ones. The difference between running in this image and the GitHub Ubuntu runner appear to be:

- Umasks are both `0022`
- Processes are running as user
  - `runner` on GitHub
  - `linuxbrew` in the Homebrew image
- The Prefix/Cellar are:
  - `drwxr-xr-x+` and owner/group runner/docker on GitHub
  - `drwxr-xr-x` and owner/group by linuxbrew/linuxbrew  in the Homebrew image

So I'm guessing it's something with not running as `linuxbrew` or the extended permissions of the prefix/cellar on the GitHub runner?
If anyone's interested in the gory details of the the workflow outputs, they're [in this PR's checks](https://github.com/drbenmorgan/homebrew-geant4/pull/6)

I don't know if use of `ghcr.io/homebrew/ubuntu22.04:master` is recommended for use outside of Homebrew's own testing infrastructure, but I'd be happy to use it here if that's the recommended/best solution! Otherwise, it looks like I need to change user/perms of the Homebrew prefix/cellar on the Github ubuntu runner?
